src/Io/Transaction.php

```diff
@@ -255,25 +256,36 @@ private function onResponseRedirect(ResponseInterface $response, RequestInterfac
     /**
      * @param RequestInterface $request
      * @param UriInterface $location
+     * @param int $statusCode
      * @return RequestInterface
      */
-    private function makeRedirectRequest(RequestInterface $request, UriInterface $location)
+    private function makeRedirectRequest(RequestInterface $request, UriInterface $location, $statusCode)
     {
         $originalHost = $request->getUri()->getHost();
-        $request = $request
-            ->withoutHeader('Host')
-            ->withoutHeader('Content-Type')
-            ->withoutHeader('Content-Length');
+        $request = $request->withoutHeader('Host');
+
+        if ($statusCode !== 307 || $statusCode !== 308) {
+            $request = $request
+                ->withoutHeader('Content-Type')
+                ->withoutHeader('Content-Length');
+        }
 
         // Remove authorization if changing hostnames (but not if just changing ports or protocols).
         if ($location->getHost() !== $originalHost) {
             $request = $request->withoutHeader('Authorization');
         }
 
-        // naïve approach..
-        $method = ($request->getMethod() === 'HEAD') ? 'HEAD' : 'GET';
+        $body = null;
+        if ($statusCode === 307 || $statusCode === 308) {
+            $method = $request->getMethod();
+            if ($request->getBody() instanceof ReadableStreamInterface) {throw new CantRedirectStreamException();}
```

Comment:
Maybe it's enough to throw a `RuntimeException` at this point, then we wouldn't need a new Exception class for this. The message could be something like "Unable to redirect request with streaming body". I think this should be enough to prevent our streaming body problem.